src/Tap.sol

```diff
@@ -75,21 +87,27 @@ contract Tap {
         require(now == rho, "Tap/rho-not-updated");
         if (what == "flight") flight = data;
         else revert("Tap/file-unrecognized-param");
+
+        emit File(what, data);
     }
     function file(bytes32 what, uint256 data) external auth {
         require(now == rho, "Tap/rho-not-updated");
         if (what == "rate") rate = data;
         else revert("Tap/file-unrecognized-param");
+
+        emit File(what, data);
     }
 
+    // --- External ---
     function pump() external stoppable {
-        require(now >= rho, "Tap/invalid-now");
-        uint256 wad = mul(now - rho, rate);
-        if (wad > 0) {
-            vat.suck(address(vow), address(this), mul(wad, RAY));
-            daiJoin.exit(address(this), wad);
-            keg.pour(flight, wad);
-        }
+        require(now > rho, "Tap/invalid-now");
+        uint256 wad = mul(now - rho, rate);  // TODO How is that working? We clearly need to increase the test coverage
         rho = now;
+
+        require(wad > 0, "Tap/invalid-wad");
+
+        vat.suck(address(vow), address(this), mul(wad, RAY));
+        daiJoin.exit(address(this), wad); // TODO need an escape way dai can be locked for ever. like cage() from FlapTap when it will work.
```

Comment:
The Dai does not sit here. It is immediately forwarded to the Keg which forwards to the end payment targets so no need for `cage()`.

Reply:
You alright in your current implementation of the keg we transfer from here as soon as we pour, but in theory it can stay there if the keg implementation change or if for some reason the keg doesn't transfer all the dai.

Reply:
Either you change the implementation to send the day to keg than it is "keg problem" or you can do something similar to cage() where you check and send back the dai to the vat where it can be manipulated.

Reply:
I see what you're saying, but I think we are okay in assuming the DAI going to the keg is gone forever as that is the case right now. If we do change the structure of the keg in the future then we can just replace the FlapTap at that point as well. Currently the cage() on the keg would be empty.

Reply:
It is, I think a design issue that create this sort of strange thing. I will think about it and push a separate PR if I can find something.
However this one is resolved, I removed the comment from the PR.